### DISCUSSION TITLE
Add OAuth/OpenID profile filter

### DISCUSSION THREAD

**Discussion by rijkvanzanten** [directus/directus#22972] [Feature Requests]
With the newly added OAuth and OpenID drivers, it's possible to automatically create the user if they don't exist in the platform yet. However, this is currently an all-or-nothing setup. Useful for private user directories, like auth0 or Okta, but less ideal for big open platforms like Facebook or Google. 

We should add a `AUTH_<PROVIDER>_PUBLIC_REGISTRATION_FILTER` that takes in [a filter rules object](https://docs.directus.io/reference/filter-rules/#filter-rules) to allow for specific external providers to have access. This could be something like

```json
{
  "email": {
    "_ends_with": "@directus.io"
  }
}
```

or

```json
{
  "profile": {
    "username": {
      "_in": ["@rijkvanzanten", "@benhaynes"]
    }
  }
}
```

**Comment by joselcvarela** [comment]
I believe you had removed this commit https://github.com/directus/directus/pull/8660/commits/33dd4b3c579304cf6e46168a46f866d4ff9497e6
 in order to have a more fine grained control, instead of just domains, am I right?

**Comment by adanielyan** [comment]
> With the newly added OAuth and OpenID drivers, it's possible to automatically create the user if they don't exist in the platform yet.

Is this feature documented anywhere? How do I enable the automatic user creation?

**Comment by adanielyan** [comment]
Never mind, I think I just found it in Environment variables: `AUTH_<PROVIDER>_ALLOW_PUBLIC_REGISTRATION`.

**Comment by aidenfoxx** [comment]
> I believe you had removed this commit [33dd4b3](https://github.com/directus/directus/commit/33dd4b3c579304cf6e46168a46f866d4ff9497e6) in order to have a more fine grained control, instead of just domains, am I right?

Yup. That is the plan 😄

**Comment by jamescammarano** [comment]
> > I believe you had removed this commit [33dd4b3](https://github.com/directus/directus/commit/33dd4b3c579304cf6e46168a46f866d4ff9497e6) in order to have a more fine grained control, instead of just domains, am I right?
> 
> Yup. That is the plan 😄

Are you working on this? If not could we talk more about the plan?

**Comment by aidenfoxx** [comment]
> > > I believe you had removed this commit [33dd4b3](https://github.com/directus/directus/commit/33dd4b3c579304cf6e46168a46f866d4ff9497e6) in order to have a more fine grained control, instead of just domains, am I right?
> > 
> > 
> > Yup. That is the plan 😄
> 
> Are you working on this? If not could we talk more about the plan?

This is on @rijkvanzanten. If I had my way it would be using the email domain filter I implemented 😂

**Comment by rijkvanzanten** [comment]
> Are you working on this? If not could we talk more about the plan?

@jaycammarano The rough plan is:

- When you register through an OAuth/OpenID provider, you get a profile with information
- The information is different per provider, and has an unknown type signature
- We'd want to have a way to only allow people to register who have a specific _thing_ in that profile information

In order to do that, we need to configure what to check for and filter against, which we can do in a "public registration filter" environment variable. We already have a filtering setup (the filer rules), and [a util that can check any arbitrary object against those rules](https://github.com/directus/directus/blob/main/packages/shared/src/utils/validate-payload.ts), so we should be able to add an if-statement in the registration part of the OAuth/OpenID/LDAP providers that checks if the payload we received from the provider validates against the custom rules configured in the env var. This would allow the user to only allow dynamic registration based on a known flag on "the other end", for example "role = xyz" or "email endswidth @directus.io" etc
